src/main/java/graphql/schema/validation/TypeAndFieldRule.java

```diff
@@ -0,0 +1,274 @@
+        assertTypeName(type.getName(), errorCollector);
+
+        List<GraphQLFieldDefinition> fieldDefinitions = type.getFieldDefinitions();
+        if (fieldDefinitions == null || fieldDefinitions.size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.ObjectTypeLackOfFieldError, String.format("\"%s\" must define one or more fields.", type.getName()));
+            errorCollector.addError(validationError);
+            return;
+        }
+
+        if (fieldDefinitions != null && fieldDefinitions.size() != 0) {
+            for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions) {
+                validateFieldDefinition(type.getName(), fieldDefinition, errorCollector);
+            }
+        }
+    }
+
+    private void validateInterface(GraphQLInterfaceType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+        List<GraphQLFieldDefinition> fieldDefinitions = type.getFieldDefinitions();
+        if (fieldDefinitions == null || fieldDefinitions.size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.InterfaceLackOfFieldError, type.getName() + " must define one or more fields.");
+            errorCollector.addError(validationError);
+        }
+
+        if (fieldDefinitions != null && fieldDefinitions.size() != 0) {
+            for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions) {
+                validateFieldDefinition(type.getName(), fieldDefinition, errorCollector);
+            }
+        }
+    }
+
+    private void validateInputObject(GraphQLInputObjectType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+        if (type.getFieldDefinitions() == null || type.getFieldDefinitions().size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.InputObjectTypeLackOfFieldError, type.getName() + " must define one or more fields.");
+            errorCollector.addError(validationError);
+        }
+
+    }
+
+    private void validateUnion(GraphQLUnionType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+
+        UnionTypeDefinition definition = type.getDefinition();
+        if (definition == null) {
+            return;
+        }
+
+        List<Type> memberTypes = definition.getMemberTypes();
+        if (memberTypes == null || memberTypes.size() == 0) {
+            SchemaValidationError validationError =
+                    new SchemaValidationError(SchemaValidationErrorType.UnionTypeLackOfTypeError, String.format("Union type \"%s\" must include one or more unique member types.", type.getName()));
+            errorCollector.addError(validationError);
+        }
+
+        List<String> typeNames = new ArrayList<>();
```

Comment:
I think a Set is better here as track just the names we already saw.

Reply:
Thanks for your suggestion.